### PR TITLE
3192 - Rip and replace zeroclipboard with clipboardjs

### PR DESCRIPTION
**Pull request by millayr**
Removing flash-based zeroclipboard.
Created generic React Copy component using clipboard.js.

Relevant JIRA ticket:  https://issues.apache.org/jira/browse/COUCHDB-3192
Relevant Cloudant ticket:  https://cloudant.fogbugz.com/f/cases/76320/

### CODE REVIEW

**Comment by millayr** [comment]
Zeroclipboard removed and new Copy Component in place.  Still need to write tests...hopefully this week.

File: app/addons/fauxton/assets/less/components.less [hunk 1]
```diff
@@ -44,6 +40,7 @@
 
 }
 
-a.clipboard-copy-element:hover {
-  text-decoration: none;
-}
+button.clipboard-copy-element {
+  background: transparent;
+  border: 0;
+}
```
Comment (robertkowalski) [inline app/addons/fauxton/assets/less/components.less]: trailing whitespace at EOF missing

you can set your editor to automatically add them

File: app/addons/fauxton/components.react.jsx [hunk 2]
```diff
@@ -14,180 +14,126 @@ import app from "../../app";
...
-
-  onClick: function (event) {
-    event.preventDefault();
+    // this one function is all the lib offers. It parses the entire page and pretty-prints anything with
+    // a .prettyprint class; only executes on an element once
+    prettyPrint();
   },
 
   render: function () {
+    var code = JSON.stringify(this.props.code, null, " ");
```
Comment (robertkowalski) [inline app/addons/fauxton/components.react.jsx]: const

File: app/addons/fauxton/components.react.jsx [hunk 3]
```diff
@@ -14,180 +14,126 @@ import app from "../../app";
...
+      if (done) {
+        done(false);
+      }
+    }.bind(this));
   },
 
   render: function () {
-    var code = JSON.stringify(this.props.code, null, " ");
+    var styleSpec = this.state.show ? {"display": "block", "opacity": 1} :  {"display": "none", "opacity": 0};
+    var classSpec = this.props.className || "";
```
Comment (robertkowalski) [inline app/addons/fauxton/components.react.jsx]: const

File: app/addons/fauxton/components.react.jsx [hunk 4]
```diff
@@ -14,180 +14,126 @@ import app from "../../app";
...
 
-  getCopyIcon: function () {
-    if (!this.props.showCopyIcon) {
-      return null;
+  setVisible: function (visible, done) {
+    if (this.state.show && !visible) {
+      this.hide(done);
+    } else if (!this.state.show && visible) {
+      this.show(done);
     }
```
Comment (robertkowalski) [inline app/addons/fauxton/components.react.jsx]: you can use early returns here, the else if nesting is a bit hard to read

File: app/addons/fauxton/components.react.jsx [hunk 5]
```diff
@@ -14,180 +14,126 @@ import app from "../../app";
...
-      this.clipboard.on('copy', function () {
-        this.props.onClipBoardClick();
-      }.bind(this));
-    }.bind(this));
+  toggle: function (done) {
+    if (this.state.show) {
+      this.hide(done);
+    } else {
+      this.show(done);
+    }
```
Comment (robertkowalski) [inline app/addons/fauxton/components.react.jsx]: use an early return instead of nesting with else.

File: app/addons/fauxton/components.react.jsx [hunk 6]
```diff
@@ -14,180 +14,126 @@ import app from "../../app";
...
   }
 });
 
-// use like this:
-//  <ComponentsReact.ClipboardWithTextField textToCopy={yourText} uniqueKey={someUniqueValue}>
-//  </ComponentsReact.ClipboardWithTextField>
-// pass in the text and a unique key, the key has to be unique or you'll get a warning
-var ClipboardWithTextField = React.createClass({
+var _NextTrayInternalId = 0;
+var Tray = React.createClass({
```
Comment (robertkowalski) [inline app/addons/fauxton/components.react.jsx]: tray? i think this is deleted code? maybe from a wrongly resolved conflict?

File: app/addons/fauxton/components.react.jsx [hunk 7]
```diff
@@ -14,180 +14,126 @@ import app from "../../app";
...
-var Clipboard = React.createClass({
-  propTypes: function () {
-    return {
-      text: React.PropTypes.string.isRequired,
-      displayType: React.PropTypes.string.oneOf(['icon', 'text'])
-    };
-  },
 
+// formats a block of code and pretty-prints it in the page. Currently uses the prettyPrint plugin
+var CodeFormat = React.createClass({
```
Comment (robertkowalski) [inline app/addons/fauxton/components.react.jsx]: how does this come into the diff?

**Comment by millayr** [comment]
Heh, yeah a bit of a git issue for app/addons/fauxton/components.react.jsx.  The only change I made was mass deletion of ZeroClipboard components but git is picking up old code as new additions.  I'll address your comments nevertheless as it's probably worthwhile to get that code up to date.  @robertkowalski

**Comment by robertkowalski** [comment]
good work ,+1

**Comment by robertkowalski** [comment]
squashed and merged as 2e770d66b2b849a60468ad390dfed57da54561a9

**Pull request closed without merging**